// @context-labs/models/.dist/mongo/Bot.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const GPTModels_1 = require("../enums/GPTModels");
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
const EmbedConfiguration = {
    enabled: {
        type: Boolean,
        required: true,
        default: true,
    },
    theme: {
        type: String,
        required: true,
        default: 'light',
    },
    type: {
        type: String,
        required: true,
        default: 'launcher',
    },
    position: {
        type: String,
        required: true,
        default: 'bottom-right',
    },
    background: {
        type: String,
        required: true,
        default: '#000000',
    },
    color: {
        type: String,
        required: true,
        default: '#FFFFFF',
    },
};
const TrackingConfiguration = {
    utms: [
        {
            type: String,
            required: false,
            default: [],
        },
    ],
};
const PromptConfiguration = {
    topic: {
        type: String,
        required: false,
        default: null,
    },
    context: {
        type: String,
        required: false,
        default: null,
    },
};
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
        default: null,
    },
    sourceIds: [{
            type: String,
            required: false,
            default: [],
        }],
    title: {
        type: String,
        required: true,
        default: null,
    },
    description: {
        type: String,
        required: false,
        default: null,
    },
    imageFileId: {
        type: String,
        required: false,
        default: null,
    },
    brandColor: {
        type: String,
        required: false,
        default: '#000000',
    },
    website: {
        type: String,
        required: false,
        default: null,
    },
    faqs: [{
            type: String,
            required: false,
            default: [],
        }],
    model: {
        type: String,
        required: false,
        default: GPTModels_1.zGPTModels.enum.GPT35Turbo,
    },
    archived: {
        type: Boolean,
        required: false,
        default: false,
    },
    embed: EmbedConfiguration,
    tracking: TrackingConfiguration,
    prompt: PromptConfiguration,
    meta: Metadata_mongo_1.default,
};
