// @context-labs/models/.dist/mongo/Document.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
const SRTLine = {
    id: {
        type: String,
        required: false,
    },
    startTime: {
        type: String,
        required: false,
    },
    startSeconds: {
        type: Number,
        required: false,
    },
    endTime: {
        type: String,
        required: false,
    },
    endSeconds: {
        type: Number,
        required: false,
    },
    text: {
        type: String,
        required: false,
        default: '',
    },
    pageNumber: {
        type: Number,
        required: false,
    },
};
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
    },
    sourceId: {
        type: String,
        required: true,
    },
    syncId: {
        type: String,
        required: true,
    },
    title: {
        type: String,
        required: true,
        default: null,
    },
    description: {
        type: String,
        required: true,
        default: null,
    },
    releasedAt: {
        type: Number,
        required: true,
        default: null,
    },
    type: {
        type: String,
        required: true,
    },
    url: {
        type: String,
        required: true,
    },
    domain: {
        type: String,
        required: false,
        default: null,
    },
    lang: {
        type: String,
        required: false,
        default: null,
    },
    views: {
        type: Number,
        required: false,
        default: null,
    },
    audioFileId: {
        type: String,
        required: false,
        default: null,
    },
    imageFileId: {
        type: String,
        required: false,
        default: null,
    },
    predictionId: {
        type: String,
        required: false,
        default: null,
    },
    raw: {
        type: String,
        required: false,
        default: null,
    },
    srt: [SRTLine],
    importStatus: {
        type: String,
        required: false,
        default: null,
    },
    archived: {
        type: Boolean,
        required: false,
        default: false,
    },
    deleted: {
        type: Boolean,
        required: false,
        default: false,
    },
    meta: Metadata_mongo_1.default,
};
