// @context-labs/models/.dist/mongo/Invoice.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
const Usage = {
    credits: {
        type: Number,
        required: true,
        default: 0,
    },
    documents: {
        type: Number,
        required: true,
        default: 0,
    },
    syncs: {
        type: Number,
        required: true,
        default: 0,
    },
};
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
    },
    subscriptionId: {
        type: String,
        required: true,
    },
    stripeCustomerId: {
        type: String,
        required: true,
    },
    stripeSubscriptionId: {
        type: String,
        required: false,
        default: null,
    },
    stripeInvoiceId: {
        type: String,
        required: false,
        default: null,
    },
    stripeProductId: {
        type: String,
        required: false,
        default: null,
    },
    stripePriceId: {
        type: String,
        required: false,
        default: null,
    },
    status: {
        type: String,
        required: true,
    },
    usage: Usage,
    usageLimits: Usage,
    meta: Metadata_mongo_1.default,
};
