// @context-labs/models/.dist/mongo/Query.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Geolocation_mongo_1 = __importDefault(require("./Geolocation.mongo"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    distinctId: {
        type: String,
        required: false,
        default: null,
    },
    organizationId: {
        type: String,
        required: true,
    },
    botId: {
        type: String,
        required: false,
    },
    sourceIds: [{
            type: String,
            required: true,
        }],
    snippetIds: [{
            type: String,
            required: false,
            default: [],
        }],
    input: {
        type: String,
        required: false,
        default: null,
    },
    output: {
        type: String,
        required: false,
        default: null,
    },
    helpful: {
        type: Boolean,
        required: false,
        default: null,
    },
    mark: {
        type: Boolean,
        required: false,
        default: null,
    },
    edits: [{
            _id: {
                type: String,
                default: shortid_1.default.generate,
                required: false,
            },
            value: {
                type: String,
                required: true,
            },
            meta: Metadata_mongo_1.default,
        }],
    model: {
        type: String,
        required: false,
        default: null,
    },
    embedding: [{
            type: Number,
            required: false,
            default: [],
        }],
    type: {
        type: String,
        required: true,
    },
    origin: {
        channel: {
            type: String,
            required: false,
            default: null,
        },
        originId: {
            type: String,
            required: false,
            default: null,
        },
        originName: {
            type: String,
            required: false,
            default: null,
        },
        originChannelId: {
            type: String,
            required: false,
            default: null,
        },
        originChannelName: {
            type: String,
            required: false,
            default: null,
        },
    },
    cachedQueryId: {
        type: String,
        required: false,
        default: null,
    },
    lang: {
        type: String,
        required: false,
        default: null,
    },
    geolocation: Geolocation_mongo_1.default,
    meta: Metadata_mongo_1.default,
};
