// @context-labs/models/.dist/mongo/Slack.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
const Team = {
    id: {
        type: String,
        required: true,
    },
    name: {
        type: String,
        required: true,
    },
};
// The stripe bot becomes a user in the slack workspace
const AuthedUser = {
    id: {
        type: String,
        required: true,
    },
    scope: {
        type: String,
    },
    accessToken: {
        type: String,
    },
};
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    access_token: {
        type: String,
        required: true,
        unique: true,
    },
    scope: {
        type: String,
        required: false,
    },
    botId: {
        type: String,
        required: false,
    },
    appId: {
        type: String,
        required: false,
    },
    team: Team,
    meta: Metadata_mongo_1.default,
    authedUser: AuthedUser,
};
