// @context-labs/models/.dist/mongo/Snippet.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
    },
    sourceId: {
        type: String,
        required: true,
    },
    syncId: {
        type: String,
        required: true,
    },
    documentId: {
        type: String,
        required: true,
    },
    title: {
        type: String,
        required: true,
        default: null,
    },
    text: {
        type: String,
        required: true,
        default: null,
    },
    type: {
        type: String,
        required: true,
    },
    url: {
        type: String,
        required: true,
    },
    domain: {
        type: String,
        required: false,
        default: null,
    },
    lang: {
        type: String,
        required: false,
        default: null,
    },
    hash: {
        type: String,
        required: true,
        default: null,
        unique: true,
    },
    index: {
        type: Number,
        required: true,
    },
    startTime: {
        type: String,
        required: true,
        default: null,
    },
    startSeconds: {
        type: Number,
        required: true,
    },
    endTime: {
        type: String,
        required: true,
        default: null,
    },
    endSeconds: {
        type: Number,
        required: true,
    },
    releasedAt: {
        type: Number,
        required: true,
    },
    views: {
        type: Number,
        required: false,
        default: null,
    },
    pageNumber: {
        type: Number,
        required: false,
        default: null,
    },
    meta: Metadata_mongo_1.default,
};
