// @context-labs/models/.dist/mongo/Source.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
        default: null,
    },
    syncId: {
        type: String,
        required: false,
        default: null,
    },
    title: {
        type: String,
        required: true,
        default: null,
    },
    description: {
        type: String,
        required: false,
        default: null,
    },
    imageFileId: {
        type: String,
        required: false,
        default: null,
    },
    type: {
        type: String,
        required: true,
    },
    importUrl: {
        type: String,
        required: false,
        default: null,
    },
    transcribe: {
        type: Boolean,
        required: false,
        default: false,
    },
    regex: {
        type: String,
        required: false,
        default: null,
    },
    durationSeconds: {
        type: Number,
        required: false,
        default: 0,
    },
    documentCount: {
        type: Number,
        required: false,
        default: 0,
    },
    archived: {
        type: Boolean,
        required: false,
        default: false,
    },
    meta: Metadata_mongo_1.default,
};
