// @context-labs/models/.dist/mongo/Sync.mongo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const shortid_1 = __importDefault(require("shortid"));
const Metadata_mongo_1 = __importDefault(require("./Metadata.mongo"));
exports.default = {
    _id: {
        type: String,
        required: true,
        default: shortid_1.default.generate,
    },
    organizationId: {
        type: String,
        required: true,
    },
    sourceId: {
        type: String,
        required: true,
    },
    type: {
        type: String,
        required: true,
        default: null,
    },
    status: {
        type: String,
        required: true,
        default: null,
    },
    message: {
        type: String,
        required: false,
        default: null,
    },
    info: {
        total: {
            type: Number,
            required: true,
        },
        new: {
            type: Number,
            required: true,
        },
        existing: {
            type: Number,
            required: true,
        },
        removed: {
            type: Number,
            required: true,
        },
        pending: {
            type: Number,
            required: true,
        },
        running: {
            type: Number,
            required: true,
        },
        completed: {
            type: Number,
            required: true,
        },
        failed: {
            type: Number,
            required: true,
        },
    },
    meta: Metadata_mongo_1.default,
};
