// @context-labs/models/src/utils/DocumentUtil.ts
import { SourceTypes, zSourceTypes } from '../services/SourceService';
import {
  Document, DocumentSummary, DocumentTypes, zDocumentTypes,
} from '../services/DocumentService';

const convertToSummary = (document: Document): DocumentSummary => {
  return {
    _id: document._id,
    syncId: document.syncId,
    organizationId: document.organizationId,
    sourceId: document.sourceId,
    title: document.title,
    description: document.description ?? '',
    type: document.type,
    url: document.url,
    audioFileId: document.audioFileId,
    imageFileId: document.imageFileId,
    releasedAt: document.releasedAt,
    importStatus: document.importStatus,
    meta: document.meta,
  };
};

const typeFromSourceType = (sourceType: SourceTypes): DocumentTypes => {
  return {
    [zSourceTypes.Enum.YouTube]: zDocumentTypes.Enum.YouTube,
    [zSourceTypes.Enum.RSS]: zDocumentTypes.Enum.RSS,
    [zSourceTypes.Enum.SiteMap]: zDocumentTypes.Enum.WebPage,
    [zSourceTypes.Enum.Files]: zDocumentTypes.Enum.File,
    [zSourceTypes.Enum.WARC]: zDocumentTypes.Enum.WebPage,
  }[sourceType];
};

const DocumentUtil = {
  convertToSummary,
  typeFromSourceType,
};

export default DocumentUtil;
