// @context-labs/models/.dist/utils/DocumentUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SourceService_1 = require("../services/SourceService");
const DocumentService_1 = require("../services/DocumentService");
const convertToSummary = (document) => {
    var _a;
    return {
        _id: document._id,
        syncId: document.syncId,
        organizationId: document.organizationId,
        sourceId: document.sourceId,
        title: document.title,
        description: (_a = document.description) !== null && _a !== void 0 ? _a : '',
        type: document.type,
        url: document.url,
        audioFileId: document.audioFileId,
        imageFileId: document.imageFileId,
        releasedAt: document.releasedAt,
        importStatus: document.importStatus,
        meta: document.meta,
    };
};
const typeFromSourceType = (sourceType) => {
    return {
        [SourceService_1.zSourceTypes.Enum.YouTube]: DocumentService_1.zDocumentTypes.Enum.YouTube,
        [SourceService_1.zSourceTypes.Enum.RSS]: DocumentService_1.zDocumentTypes.Enum.RSS,
        [SourceService_1.zSourceTypes.Enum.SiteMap]: DocumentService_1.zDocumentTypes.Enum.WebPage,
        [SourceService_1.zSourceTypes.Enum.Files]: DocumentService_1.zDocumentTypes.Enum.File,
        [SourceService_1.zSourceTypes.Enum.WARC]: DocumentService_1.zDocumentTypes.Enum.WebPage,
    }[sourceType];
};
const DocumentUtil = {
    convertToSummary,
    typeFromSourceType,
};
exports.default = DocumentUtil;
