// @context-labs/models/.dist/enums/GPTModels.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.modelCostPerQuery = exports.queryContextLength = exports.maxContextLength = exports.zGPTModels = void 0;
const zod_1 = __importDefault(require("zod"));
const types = {
    GPT35Turbo: 'gpt-3.5-turbo',
    GPT35Turbo0613: 'gpt-3.5-turbo-0613',
    GPT35Turbo16K: 'gpt-3.5-turbo-16k',
    GPT4: 'gpt-4',
    GPT40613: 'gpt-4-0613',
    GPT41106: 'gpt-4-1106-preview',
    TextEmbeddingAda002: 'text-embedding-ada-002',
};
exports.zGPTModels = zod_1.default.nativeEnum(types);
const maxContextLength = (model) => {
    switch (model) {
        case exports.zGPTModels.enum.GPT35Turbo:
            return 4096;
        case exports.zGPTModels.enum.GPT35Turbo0613:
            return 4096;
        case exports.zGPTModels.enum.GPT35Turbo16K:
            return 16384;
        case exports.zGPTModels.enum.GPT4:
            return 8192;
        case exports.zGPTModels.enum.GPT40613:
            return 8192;
        case exports.zGPTModels.enum.GPT41106:
            return 8192;
        case exports.zGPTModels.enum.TextEmbeddingAda002:
            return 8192;
        default:
            return 0;
    }
};
exports.maxContextLength = maxContextLength;
const queryContextLength = (model) => {
    switch (model) {
        case exports.zGPTModels.enum.GPT35Turbo:
            return 4000;
        case exports.zGPTModels.enum.GPT35Turbo0613:
            return 4000;
        case exports.zGPTModels.enum.GPT35Turbo16K:
            return 10000;
        case exports.zGPTModels.enum.GPT4:
            return 8000;
        case exports.zGPTModels.enum.GPT40613:
            return 8000;
        case exports.zGPTModels.enum.GPT41106:
            return 8000;
        default:
            return 0;
    }
};
exports.queryContextLength = queryContextLength;
const modelCostPerQuery = (model) => {
    switch (model) {
        case exports.zGPTModels.enum.GPT35Turbo:
            return 1;
        case exports.zGPTModels.enum.GPT35Turbo0613:
            return 1;
        case exports.zGPTModels.enum.GPT35Turbo16K:
            return 5;
        case exports.zGPTModels.enum.GPT4:
            return 20;
        case exports.zGPTModels.enum.GPT40613:
            return 20;
        case exports.zGPTModels.enum.GPT41106:
            return 20;
        default:
            return 0;
    }
};
exports.modelCostPerQuery = modelCostPerQuery;
