// @context-labs/models/.dist/services/DiscordService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBotRequest = exports.getBotParams = exports.getDiscordByServerIdRequest = exports.getDiscordByServerIdParams = exports.createDiscordRequest = exports.createDiscordParams = exports.discord = exports.discordServer = void 0;
const zod_1 = __importDefault(require("zod"));
const Metadata_1 = require("../interfaces/Metadata");
// This service doesn't respond to HTTP requests, so it doesn't need to be extend ServiceResponse
exports.discordServer = zod_1.default.object({
    id: zod_1.default.string(),
    name: zod_1.default.string(),
    registeredOn: zod_1.default.date(),
});
exports.discord = zod_1.default.object({
    _id: zod_1.default.string(),
    name: zod_1.default.string(),
    botId: zod_1.default.string(),
    server: exports.discordServer,
    meta: Metadata_1.metadata,
});
/** ******************************************************************************
*  Create Discord
******************************************************************************* */
exports.createDiscordParams = zod_1.default.object({
    name: zod_1.default.string().optional(),
    botId: zod_1.default.string(),
    server: exports.discordServer,
    meta: Metadata_1.metadata,
});
exports.createDiscordRequest = zod_1.default.object({
    params: exports.createDiscordParams,
});
/** ******************************************************************************
*  Get Discord By Server Id
******************************************************************************* */
exports.getDiscordByServerIdParams = zod_1.default.object({
    serverId: zod_1.default.string(),
});
exports.getDiscordByServerIdRequest = zod_1.default.object({
    params: exports.getDiscordByServerIdParams,
});
/** ******************************************************************************
*  Get Bot
******************************************************************************* */
exports.getBotParams = zod_1.default.object({
    botId: zod_1.default.string(),
});
exports.getBotRequest = zod_1.default.object({
    params: exports.getBotParams,
});
