// @context-labs/models/.dist/services/DiscordQueryService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDiscordQueriesByServerIdParams = exports.getDiscordQueryParams = exports.createDiscordQueryParams = exports.discordQuery = void 0;
const zod_1 = __importDefault(require("zod"));
const Metadata_1 = require("../interfaces/Metadata");
exports.discordQuery = zod_1.default.object({
    _id: zod_1.default.string(),
    discordId: zod_1.default.string(),
    organizationId: zod_1.default.string(),
    messageId: zod_1.default.string(),
    helpful: zod_1.default.boolean().optional(),
    queryId: zod_1.default.string(),
    meta: Metadata_1.metadata,
});
exports.createDiscordQueryParams = zod_1.default.object({
    discordId: zod_1.default.string(),
    organizationId: zod_1.default.string(),
    messageId: zod_1.default.string(),
    helpful: zod_1.default.boolean(),
    meta: Metadata_1.metadata,
    queryId: zod_1.default.string(),
});
exports.getDiscordQueryParams = zod_1.default.object({
    id: zod_1.default.string(),
});
exports.getDiscordQueriesByServerIdParams = zod_1.default.object({
    serverId: zod_1.default.string(),
});
