// @context-labs/models/.dist/services/EmailService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendInviteToOrganizationEmailResponse = exports.sendInviteToOrganizationEmailRequest = exports.sendInviteToOrganizationEmailParams = exports.sendSourceImportedEmailResponse = exports.sendSourceImportedEmailRequest = exports.sendSourceImportedEmailParams = exports.sendSourceCreatedEmailResponse = exports.sendSourceCreatedEmailRequest = exports.sendSourceCreatedEmailParams = void 0;
const zod_1 = __importDefault(require("zod"));
const common_1 = require("../interfaces/common");
const SourceService_1 = require("./SourceService");
const OrganizationService_1 = require("./OrganizationService");
/** ******************************************************************************
*  Send Source Created Email
******************************************************************************* */
exports.sendSourceCreatedEmailParams = zod_1.default.object({
    toAddress: zod_1.default.string(),
    source: SourceService_1.source,
});
exports.sendSourceCreatedEmailRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.sendSourceCreatedEmailParams,
}));
exports.sendSourceCreatedEmailResponse = common_1.serviceResponse.merge(zod_1.default.object({
    success: zod_1.default.boolean().optional(),
}));
/** ******************************************************************************
*  Send Source Imported Email
******************************************************************************* */
exports.sendSourceImportedEmailParams = zod_1.default.object({
    toAddress: zod_1.default.string(),
    source: SourceService_1.source,
});
exports.sendSourceImportedEmailRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.sendSourceImportedEmailParams,
}));
exports.sendSourceImportedEmailResponse = common_1.serviceResponse.merge(zod_1.default.object({
    success: zod_1.default.boolean().optional(),
}));
/** ******************************************************************************
*  Send Invite  to Organization
******************************************************************************* */
exports.sendInviteToOrganizationEmailParams = zod_1.default.object({
    toAddress: zod_1.default.string(),
    inviteeExists: zod_1.default.boolean(),
    inviterFirstName: zod_1.default.string(),
    inviterLastName: zod_1.default.string(),
    organization: OrganizationService_1.organization,
});
exports.sendInviteToOrganizationEmailRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.sendInviteToOrganizationEmailParams,
}));
exports.sendInviteToOrganizationEmailResponse = common_1.serviceResponse.merge(zod_1.default.object({
    success: zod_1.default.boolean().optional(),
}));
