// @context-labs/models/.dist/services/KeyService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listKeysResponse = exports.listKeysRequest = exports.listKeysParams = exports.getKeyResponse = exports.getKeyRequest = exports.getKeyParams = exports.archiveKeyResponse = exports.archiveKeyRequest = exports.archiveKeyParams = exports.createKeyResponse = exports.createKeyRequest = exports.createKeyParams = exports.key = void 0;
const zod_1 = __importDefault(require("zod"));
const common_1 = require("../interfaces/common");
const Metadata_1 = require("../interfaces/Metadata");
exports.key = zod_1.default.object({
    _id: zod_1.default.string(),
    organizationId: zod_1.default.string(),
    name: zod_1.default.string(),
    key: zod_1.default.string(),
    archived: zod_1.default.boolean(),
    lastUsedAt: zod_1.default.string().nullable(),
    meta: Metadata_1.metadata,
});
/** ******************************************************************************
*  Create Key
******************************************************************************* */
exports.createKeyParams = zod_1.default.object({
    name: zod_1.default.string().min(4).max(64),
});
exports.createKeyRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.createKeyParams,
}));
exports.createKeyResponse = common_1.serviceResponse.merge(zod_1.default.object({
    key: exports.key.optional(),
}));
/** ******************************************************************************
*  Archive Key
******************************************************************************* */
exports.archiveKeyParams = zod_1.default.object({
    _id: zod_1.default.string(),
});
exports.archiveKeyRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.archiveKeyParams,
}));
exports.archiveKeyResponse = common_1.serviceResponse.merge(zod_1.default.object({
    key: exports.key.optional(),
}));
/** ******************************************************************************
*  Get Key
******************************************************************************* */
exports.getKeyParams = zod_1.default.object({
    keyId: zod_1.default.string(),
});
exports.getKeyRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.getKeyParams,
}));
exports.getKeyResponse = common_1.serviceResponse.merge(zod_1.default.object({
    key: exports.key.nullable().optional(),
}));
/** ******************************************************************************
*  List Keys
******************************************************************************* */
exports.listKeysParams = zod_1.default.object({});
exports.listKeysRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.listKeysParams,
}));
exports.listKeysResponse = common_1.serviceResponse.merge(zod_1.default.object({
    keys: zod_1.default.array(exports.key).optional(),
}));
