// @context-labs/models/.dist/services/PortalService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGenerationResponse = exports.createGenerationParams = exports.semanticSearchResult = exports.semanticSearchParams = exports.zPortalActions = void 0;
const zod_1 = __importDefault(require("zod"));
const QueryService_1 = require("./QueryService");
const SnippetService_1 = require("./SnippetService");
exports.zPortalActions = zod_1.default.enum([
    'Open',
    'Close',
    'Focus',
]);
/** ******************************************************************************
*  Semantic Search
******************************************************************************* */
exports.semanticSearchParams = zod_1.default.object({
    botId: zod_1.default.string(),
    sourceIds: zod_1.default.array(zod_1.default.string()),
    syncIds: zod_1.default.array(zod_1.default.string()).optional(),
    query: zod_1.default.string(),
    topK: zod_1.default.number().optional(),
    distinctId: zod_1.default.string(),
    origin: QueryService_1.queryOrigin.nullable(),
});
exports.semanticSearchResult = zod_1.default.object({
    snippets: zod_1.default.array(SnippetService_1.snippet),
});
/** ******************************************************************************
*  Create Generation
******************************************************************************* */
exports.createGenerationParams = zod_1.default.object({
    botId: zod_1.default.string(),
    sourceIds: zod_1.default.array(zod_1.default.string()),
    query: zod_1.default.string(),
    distinctId: zod_1.default.string(),
    origin: QueryService_1.queryOrigin.optional(),
});
exports.createGenerationResponse = zod_1.default.object({
    query: QueryService_1.query.optional(),
    snippets: zod_1.default.array(SnippetService_1.snippet).optional(),
    error: zod_1.default.string().optional(),
});
