// @context-labs/models/.dist/services/SitemapService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listSitemapsResponse = exports.listSitemapsRequest = exports.listSitemapsParams = exports.getSitemapResponse = exports.getSitemapRequest = exports.getSitemapParams = exports.createSitemapResponse = exports.createSitemapRequest = exports.createSitemapParams = exports.sitemap = void 0;
const zod_1 = __importDefault(require("zod"));
const common_1 = require("../interfaces/common");
const Metadata_1 = require("../interfaces/Metadata");
exports.sitemap = zod_1.default.object({
    _id: zod_1.default.string(),
    organizationId: zod_1.default.string(),
    sitemapFileId: zod_1.default.string(),
    sitemapUrl: zod_1.default.string(),
    urls: zod_1.default.array(zod_1.default.string()),
    meta: Metadata_1.metadata,
});
/** ******************************************************************************
*  Create Sitemap
******************************************************************************* */
exports.createSitemapParams = zod_1.default.object({
    urls: zod_1.default.array(zod_1.default.string()),
});
exports.createSitemapRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.createSitemapParams,
}));
exports.createSitemapResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sitemap: exports.sitemap.optional(),
}));
/** ******************************************************************************
*  Get Sitemap
******************************************************************************* */
exports.getSitemapParams = zod_1.default.object({
    sitemapId: zod_1.default.string(),
});
exports.getSitemapRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.getSitemapParams,
}));
exports.getSitemapResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sitemap: exports.sitemap.optional(),
}));
/** ******************************************************************************
*  List Sitemaps
******************************************************************************* */
exports.listSitemapsParams = zod_1.default.object({});
exports.listSitemapsRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.listSitemapsParams,
}));
exports.listSitemapsResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sitemaps: zod_1.default.array(exports.sitemap).optional(),
}));
