// @context-labs/models/.dist/services/SyncService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listSyncsResponse = exports.listSyncsRequest = exports.listSyncsParams = exports.latestSyncResponse = exports.latestSyncRequest = exports.latestSyncParams = exports.getSyncResponse = exports.getSyncRequest = exports.getSyncParams = exports.updateSyncResponse = exports.updateSyncRequest = exports.updateSyncParams = exports.createSyncResponse = exports.createSyncRequest = exports.createSyncParams = exports.sync = exports.updateSyncInfoParams = exports.syncInfo = exports.zSyncStatus = exports.zSyncTypes = void 0;
const zod_1 = __importDefault(require("zod"));
const common_1 = require("../interfaces/common");
const Metadata_1 = require("../interfaces/Metadata");
exports.zSyncTypes = zod_1.default.enum([
    'Initial',
    'Manual',
    'Automatic',
]);
exports.zSyncStatus = zod_1.default.enum([
    'Pending',
    'Running',
    'Complete',
    'Failed',
]);
exports.syncInfo = zod_1.default.object({
    total: zod_1.default.number(),
    new: zod_1.default.number(),
    existing: zod_1.default.number(),
    removed: zod_1.default.number(),
    pending: zod_1.default.number(),
    running: zod_1.default.number(),
    completed: zod_1.default.number(),
    failed: zod_1.default.number(),
});
exports.updateSyncInfoParams = zod_1.default.object({
    total: zod_1.default.number().optional(),
    new: zod_1.default.number().optional(),
    existing: zod_1.default.number().optional(),
    removed: zod_1.default.number().optional(),
    pending: zod_1.default.number().optional(),
    running: zod_1.default.number().optional(),
    completed: zod_1.default.number().optional(),
    failed: zod_1.default.number().optional(),
});
exports.sync = zod_1.default.object({
    _id: zod_1.default.string(),
    organizationId: zod_1.default.string(),
    sourceId: zod_1.default.string(),
    type: exports.zSyncTypes,
    status: exports.zSyncStatus,
    message: zod_1.default.string().optional(),
    info: exports.syncInfo,
    meta: Metadata_1.metadata,
});
/** ******************************************************************************
*  Create Sync
******************************************************************************* */
exports.createSyncParams = zod_1.default.object({
    sourceId: zod_1.default.string(),
    type: exports.zSyncTypes,
});
exports.createSyncRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.createSyncParams,
}));
exports.createSyncResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sync: exports.sync.optional(),
}));
/** ******************************************************************************
*  Update Sync
******************************************************************************* */
exports.updateSyncParams = zod_1.default.object({
    _id: zod_1.default.string(),
    status: exports.zSyncStatus.nullable().optional(),
    message: zod_1.default.string().nullable().optional(),
    infoDiff: exports.updateSyncInfoParams.optional(),
});
exports.updateSyncRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.updateSyncParams,
}));
exports.updateSyncResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sync: exports.sync.nullable().optional(),
}));
/** ******************************************************************************
*  Get Sync
******************************************************************************* */
exports.getSyncParams = zod_1.default.object({
    syncId: zod_1.default.string(),
});
exports.getSyncRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.getSyncParams,
}));
exports.getSyncResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sync: exports.sync.nullable().optional(),
}));
/** ******************************************************************************
*  Latest Sync
******************************************************************************* */
exports.latestSyncParams = zod_1.default.object({
    sourceId: zod_1.default.string(),
});
exports.latestSyncRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.latestSyncParams,
}));
exports.latestSyncResponse = common_1.serviceResponse.merge(zod_1.default.object({
    sync: exports.sync.nullable().optional(),
}));
/** ******************************************************************************
*  List Syncs
******************************************************************************* */
exports.listSyncsParams = zod_1.default.object({
    sourceId: zod_1.default.string(),
});
exports.listSyncsRequest = common_1.authedServiceRequest.merge(zod_1.default.object({
    params: exports.listSyncsParams,
}));
exports.listSyncsResponse = common_1.serviceResponse.merge(zod_1.default.object({
    syncs: zod_1.default.array(exports.sync).optional(),
}));
