// @context-labs/models/.dist/services/TaskService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUpdateDocumentTaskResponse = exports.createUpdateDocumentTaskRequest = exports.createUpdateDocumentTaskParams = exports.createImportDocumentTaskResponse = exports.createImportDocumentTaskRequest = exports.createImportDocumentTaskParams = exports.createImportSourceTaskResponse = exports.createImportSourceTaskRequest = exports.createImportSourceTaskParams = exports.TaskTypes = void 0;
const zod_1 = __importDefault(require("zod"));
const common_1 = require("../interfaces/common");
var TaskTypes;
(function (TaskTypes) {
    TaskTypes["ImportSource"] = "Import Source";
    TaskTypes["StartImportDocument"] = "Start Import Document";
    TaskTypes["ImportDocumentTranscriptFromYouTube"] = "Import Document Transcript From YouTube";
    TaskTypes["CreateDocumentSnippets"] = "Create Document Snippets";
    TaskTypes["ImportDocumentViewsFromYouTube"] = "Import Document Views From YouTube";
    TaskTypes["ImportDocumentAudio"] = "Import Document Audio";
    TaskTypes["ImportDocumentWebPage"] = "Import Document Web Page";
    TaskTypes["ImportUnstructuredDocument"] = "Import Unstructured Document";
    TaskTypes["StartImportDocumentTranscript"] = "Start Import Document Transcript";
    TaskTypes["FinishImportDocumentTranscript"] = "Finish Import Document Transcript";
    TaskTypes["FinishImportDocument"] = "Finish Import Document";
})(TaskTypes || (exports.TaskTypes = TaskTypes = {}));
/** ******************************************************************************
*  Create Import Source
******************************************************************************* */
exports.createImportSourceTaskParams = zod_1.default.object({
    taskId: zod_1.default.string().optional(),
    syncId: zod_1.default.string().optional(),
    sourceId: zod_1.default.string().optional(),
    runAt: zod_1.default.number().nullable().optional(),
});
exports.createImportSourceTaskRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.createImportSourceTaskParams,
}));
exports.createImportSourceTaskResponse = common_1.serviceResponse.merge(zod_1.default.object({
    taskId: zod_1.default.string().optional(),
    success: zod_1.default.boolean().optional(),
}));
/** ******************************************************************************
*  Create Import Document
******************************************************************************* */
exports.createImportDocumentTaskParams = zod_1.default.object({
    taskId: zod_1.default.string().optional(),
    documentId: zod_1.default.string(),
    syncId: zod_1.default.string(),
    useReplicate: zod_1.default.boolean(),
    runAt: zod_1.default.number().optional().nullable(),
});
exports.createImportDocumentTaskRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.createImportDocumentTaskParams,
}));
exports.createImportDocumentTaskResponse = common_1.serviceResponse.merge(zod_1.default.object({
    taskId: zod_1.default.string().optional(),
    success: zod_1.default.boolean().optional(),
}));
/** ******************************************************************************
*  Create Update Document
******************************************************************************* */
exports.createUpdateDocumentTaskParams = zod_1.default.object({
    taskId: zod_1.default.string(),
    documentId: zod_1.default.string(),
    syncId: zod_1.default.string(),
    runAt: zod_1.default.number().nullable(),
});
exports.createUpdateDocumentTaskRequest = common_1.serviceRequest.merge(zod_1.default.object({
    params: exports.createUpdateDocumentTaskParams,
}));
exports.createUpdateDocumentTaskResponse = common_1.serviceResponse.merge(zod_1.default.object({
    taskId: zod_1.default.string().optional(),
    success: zod_1.default.boolean().optional(),
}));
