// @context-labs/models/.dist/interfaces/common.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serviceResponse = exports.authedServiceRequest = exports.serviceRequest = exports.authData = exports.authToken = exports.toError = exports.serviceError = void 0;
const zod_1 = __importDefault(require("zod"));
const StatusCodes_1 = __importDefault(require("../enums/StatusCodes"));
/** ******************************************************************************
* Error Handling
******************************************************************************* */
exports.serviceError = zod_1.default.object({
    message: zod_1.default.string(),
    fields: zod_1.default.record(zod_1.default.string(), zod_1.default.string()).optional(),
});
const toError = (message) => {
    return { message };
};
exports.toError = toError;
/** ******************************************************************************
*  Authentication
******************************************************************************* */
exports.authToken = zod_1.default.string();
exports.authData = zod_1.default.object({
    email: zod_1.default.string().optional(),
    userId: zod_1.default.string().optional(),
    organizationId: zod_1.default.string().optional(),
    token: exports.authToken.optional(),
});
/** ******************************************************************************
*  Request/Response
******************************************************************************* */
exports.serviceRequest = zod_1.default.object({});
exports.authedServiceRequest = zod_1.default.object({
    auth: exports.authData,
});
const statuses = Object.keys(StatusCodes_1.default);
exports.serviceResponse = zod_1.default.object({
    status: zod_1.default.enum([statuses[0] || '', ...statuses]),
    error: exports.serviceError.optional(),
});
// /** ******************************************************************************
// *  Authentication
// ******************************************************************************* */
// export interface AuthenticationData {
//   accountId: string;
//   token: string;
//   ipAddress?: string;
// }
// export interface AuthenticatedRequest extends IRequest {
//   auth: AuthenticationData;
// }
// export interface AuthenticatedOptionalRequest extends IRequest {
//   auth: Partial<AuthorizationData>;
// }
// /** ******************************************************************************
// *  Authorization
// ******************************************************************************* */
// export interface AuthorizationData extends AuthenticationData {
//   organizationId: string;
// }
// export interface AuthorizedRequest extends IRequest {
//   auth: AuthorizationData;
// }
// export interface AuthorizedOptionalRequest extends IRequest {
//   auth: Partial<AuthorizationData>;
// }
