// @context-labs/models/.dist/interfaces/Unstructured.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unstructuredSnippet = exports.unstructuredElement = exports.zTextTypes = exports.unstructuredMetadata = void 0;
const zod_1 = __importDefault(require("zod"));
exports.unstructuredMetadata = zod_1.default.object({
    page_number: zod_1.default.number(),
    filename: zod_1.default.string(),
});
exports.zTextTypes = zod_1.default.enum([
    'Title',
    'ListItem',
    'NarrativeText',
    'UncategorizedText',
    'FigureCaption',
    'Body',
    'Header',
    'Footer',
]);
exports.unstructuredElement = zod_1.default.object({
    element_id: zod_1.default.string(),
    text: zod_1.default.string(),
    type: exports.zTextTypes,
    metadata: exports.unstructuredMetadata,
});
exports.unstructuredSnippet = zod_1.default.object({
    id: zod_1.default.string(),
    text: zod_1.default.string(),
    pageNumber: zod_1.default.number(),
});
