// @context-labs/models/src/mongo/Invoice.mongo.ts
import shortid from 'shortid';
import Metadata from './Metadata.mongo';

const Usage = {
  credits: {
    type: Number,
    required: true,
    default: 0,
  },
  documents: {
    type: Number,
    required: true,
    default: 0,
  },
  syncs: {
    type: Number,
    required: true,
    default: 0,
  },
};

export default {
  _id: {
    type: String,
    required: true,
    default: shortid.generate,
  },
  organizationId: {
    type: String,
    required: true,
  },
  subscriptionId: {
    type: String,
    required: true,
  },
  stripeCustomerId: {
    type: String,
    required: true,
  },
  stripeSubscriptionId: {
    type: String,
    required: false,
    default: null,
  },
  stripeInvoiceId: {
    type: String,
    required: false,
    default: null,
  },
  stripeProductId: {
    type: String,
    required: false,
    default: null,
  },
  stripePriceId: {
    type: String,
    required: false,
    default: null,
  },
  status: {
    type: String,
    required: true,
  },
  usage: Usage,
  usageLimits: Usage,
  meta: Metadata,
};
