// @context-labs/models/src/mongo/Snippet.mongo.ts
import shortid from 'shortid';
import Metadata from './Metadata.mongo';

export default {
  _id: {
    type: String,
    required: true,
    default: shortid.generate,
  },
  organizationId: {
    type: String,
    required: true,
  },
  sourceId: {
    type: String,
    required: true,
  },
  syncId: {
    type: String,
    required: true,
  },
  documentId: {
    type: String,
    required: true,
  },
  title: {
    type: String,
    required: true,
    default: null,
  },
  text: {
    type: String,
    required: true,
    default: null,
  },
  type: {
    type: String,
    required: true,
  },
  url: {
    type: String,
    required: true,
  },
  domain: {
    type: String,
    required: false,
    default: null,
  },
  lang: {
    type: String,
    required: false,
    default: null,
  },
  hash: {
    type: String,
    required: true,
    default: null,
    unique: true,
  },
  index: {
    type: Number,
    required: true,
  },
  startTime: {
    type: String,
    required: true,
    default: null,
  },
  startSeconds: {
    type: Number,
    required: true,
  },
  endTime: {
    type: String,
    required: true,
    default: null,
  },
  endSeconds: {
    type: Number,
    required: true,
  },
  releasedAt: {
    type: Number,
    required: true,
  },
  views: {
    type: Number,
    required: false,
    default: null,
  },
  pageNumber: {
    type: Number,
    required: false,
    default: null,
  },
  meta: Metadata,
};
