// @context-labs/models/src/mongo/Source.mongo.ts
import shortid from 'shortid';
import Metadata from './Metadata.mongo';

export default {
  _id: {
    type: String,
    required: true,
    default: shortid.generate,
  },
  organizationId: {
    type: String,
    required: true,
    default: null,
  },
  syncId: {
    type: String,
    required: false,
    default: null,
  },
  title: {
    type: String,
    required: true,
    default: null,
  },
  description: {
    type: String,
    required: false,
    default: null,
  },
  imageFileId: {
    type: String,
    required: false,
    default: null,
  },
  type: {
    type: String,
    required: true,
  },
  importUrl: {
    type: String,
    required: false,
    default: null,
  },
  transcribe: {
    type: Boolean,
    required: false,
    default: false,
  },
  regex: {
    type: String,
    required: false,
    default: null,
  },
  durationSeconds: {
    type: Number,
    required: false,
    default: 0,
  },
  documentCount: {
    type: Number,
    required: false,
    default: 0,
  },
  archived: {
    type: Boolean,
    required: false,
    default: false,
  },
  meta: Metadata,
};
