// @context-labs/models/src/mongo/Subscription.mongo.ts
import shortid from 'shortid';
import Metadata from './Metadata.mongo';

const Usage = {
  credits: {
    type: Number,
    required: true,
    default: 0,
  },
  documents: {
    type: Number,
    required: true,
    default: 0,
  },
  syncs: {
    type: Number,
    required: true,
    default: 0,
  },
};

export default {
  _id: {
    type: String,
    required: true,
    default: shortid.generate,
  },
  organizationId: {
    type: String,
    required: true,
  },
  stripeCustomerId: {
    type: String,
    required: true,
    default: null,
  },
  stripeSubscriptionId: {
    type: String,
    required: false,
    default: null,
  },
  stripeProductId: {
    type: String,
    required: false,
    default: null,
  },
  stripePriceId: {
    type: String,
    required: false,
    default: null,
  },
  status: {
    type: String,
    required: true,
  },
  upgrade: {
    priceId: {
      type: String,
      required: false,
      default: null,
    },
    productId: {
      type: String,
      required: false,
      default: null,
    },
    botId: {
      type: String,
      required: false,
      default: null,
    },
    sourceIds: {
      type: [String],
      required: false,
      default: null,
    },
  },
  usageLimits: Usage,
  meta: Metadata,
};
