// @context-labs/models/src/mongo/Sync.mongo.ts
import shortid from 'shortid';
import Metadata from './Metadata.mongo';

export default {
  _id: {
    type: String,
    required: true,
    default: shortid.generate,
  },
  organizationId: {
    type: String,
    required: true,
  },
  sourceId: {
    type: String,
    required: true,
  },
  type: {
    type: String,
    required: true,
    default: null,
  },
  status: {
    type: String,
    required: true,
    default: null,
  },
  message: {
    type: String,
    required: false,
    default: null,
  },
  info: {
    total: {
      type: Number,
      required: true,
    },
    new: {
      type: Number,
      required: true,
    },
    existing: {
      type: Number,
      required: true,
    },
    removed: {
      type: Number,
      required: true,
    },
    pending: {
      type: Number,
      required: true,
    },
    running: {
      type: Number,
      required: true,
    },
    completed: {
      type: Number,
      required: true,
    },
    failed: {
      type: Number,
      required: true,
    },
  },

  meta: Metadata,
};
